File: src/services/headingExtractor.ts
import * as cheerio from 'cheerio';

export interface HeadingItem {
  id: string;
  tag: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
  level: number; // 1 to 6
  text: string;
  elementId?: string;
  className?: string;
  isSkippedLevel: boolean; // True if this heading skips a level (e.g. H1 -> H3)
  parentTag?: string;
}

export interface HeadingTreeNode {
  heading: HeadingItem;
  children: HeadingTreeNode[];
}

export interface HeadingExtractionResult {
  totalHeadings: number;
  counts: {
    h1: number;
    h2: number;
    h3: number;
    h4: number;
    h5: number;
    h6: number;
  };
  issues: {
    missingH1: boolean;
    multipleH1: boolean;
    hasSkippedLevels: boolean;
    emptyHeadingsCount: number;
  };
  flatHeadings: HeadingItem[];
  tree: HeadingTreeNode[];
}

export function extractHeadingsFromHtml(html: string): HeadingExtractionResult {
  const $ = cheerio.load(html || '');
  const flatHeadings: HeadingItem[] = [];

  const counts = {
    h1: 0,
    h2: 0,
    h3: 0,
    h4: 0,
    h5: 0,
    h6: 0
  };

  let emptyHeadingsCount = 0;
  let hasSkippedLevels = false;
  let previousLevel = 0;

  $('h1, h2, h3, h4, h5, h6').each((_, el) => {
    const tagName = el.tagName.toLowerCase() as 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
    const level = parseInt(tagName.replace('h', ''), 10);
    const text = $(el).text().trim();
    const elementId = $(el).attr('id')?.trim();
    const className = $(el).attr('class')?.trim();

    if (!text) {
      emptyHeadingsCount++;
    }

    counts[tagName]++;

    // Check level skipping (e.g. going from level 1 to level 3)
    let isSkippedLevel = false;
    if (previousLevel > 0 && level > previousLevel + 1) {
      isSkippedLevel = true;
      hasSkippedLevels = true;
    }

    flatHeadings.push({
      id: `h_${flatHeadings.length + 1}`,
      tag: tagName,
      level,
      text,
      elementId,
      className,
      isSkippedLevel
    });

    previousLevel = level;
  });

  // Build Hierarchical Tree
  const tree: HeadingTreeNode[] = [];
  const stack: { level: number; node: HeadingTreeNode }[] = [];

  for (const item of flatHeadings) {
    const node: HeadingTreeNode = { heading: item, children: [] };

    while (stack.length > 0 && stack[stack.length - 1].level >= item.level) {
      stack.pop();
    }

    if (stack.length === 0) {
      tree.push(node);
    } else {
      const parent = stack[stack.length - 1].node;
      parent.children.push(node);
      item.parentTag = parent.heading.tag;
    }

    stack.push({ level: item.level, node });
  }

  return {
    totalHeadings: flatHeadings.length,
    counts,
    issues: {
      missingH1: counts.h1 === 0,
      multipleH1: counts.h1 > 1,
      hasSkippedLevels,
      emptyHeadingsCount
    },
    flatHeadings,
    tree
  };
}

File: backend/services/heading_extractor.py
from typing import Dict, Any, List, Optional
from bs4 import BeautifulSoup

class HeadingExtractor:
    """
    Extracts headings (H1-H6) from HTML and analyzes hierarchical structure,
    detecting skipped heading levels, missing/multiple H1s, and building a tree.
    """

    @staticmethod
    def extract_headings_from_html(html: str) -> Dict[str, Any]:
        soup = BeautifulSoup(html or "", "html.parser")
        headings = soup.find_all(["h1", "h2", "h3", "h4", "h5", "h6"])

        flat_headings: List[Dict[str, Any]] = []
        counts = {"h1": 0, "h2": 0, "h3": 0, "h4": 0, "h5": 0, "h6": 0}

        empty_headings_count = 0
        has_skipped_levels = False
        previous_level = 0

        for idx, el in enumerate(headings):
            tag_name = el.name.lower()
            level = int(tag_name.replace("h", ""))
            text = el.get_text().strip()
            element_id = el.get("id")
            class_name = " ".join(el.get("class")) if isinstance(el.get("class"), list) else el.get("class")

            if not text:
                empty_headings_count += 1

            counts[tag_name] += 1

            is_skipped = False
            if previous_level > 0 and level > previous_level + 1:
                is_skipped = True
                has_skipped_levels = True

            item = {
                "id": f"h_{idx + 1}",
                "tag": tag_name,
                "level": level,
                "text": text,
                "elementId": element_id,
                "className": class_name,
                "isSkippedLevel": is_skipped,
                "parentTag": None
            }
            flat_headings.append(item)
            previous_level = level

        # Build Hierarchical Tree
        tree: List[Dict[str, Any]] = []
        stack: List[Dict[str, Any]] = []

        for item in flat_headings:
            node = {"heading": item, "children": []}

            while stack and stack[-1]["level"] >= item["level"]:
                stack.pop()

            if not stack:
                tree.append(node)
            else:
                parent = stack[-1]["node"]
                parent["children"].append(node)
                item["parentTag"] = parent["heading"]["tag"]

            stack.append({"level": item["level"], "node": node})

        return {
            "totalHeadings": len(flat_headings),
            "counts": counts,
            "issues": {
                "missingH1": counts["h1"] == 0,
                "multipleH1": counts["h1"] > 1,
                "hasSkippedLevels": has_skipped_levels,
                "emptyHeadingsCount": empty_headings_count
            },
            "flatHeadings": flat_headings,
            "tree": tree
        }
